### DISCUSSION TITLE
How can i show the pred_score of each class??

### DISCUSSION THREAD

**Discussion by sinyoung-park** [open-mmlab/mmclassification#1374] [Q&A]
Hi, i'm study CNN to mmclassification in R.O.K. and i study mechainical eng. 
i make CNN code in colab follow to mmclassification tutorial.
i make the inference model too.
but the inference model return to pred_label, pred_class, pred_score when i put in image file. 
but, i want pred_score of each class.
my dataset is Customdataset.
my dataset num_classse is 5
there are my inference code.

from mmcls.apis import inference_model, init_model
import mmcv

# load image
img = mmcv.imread('/content/mmclassification/pcldata/test_set/12slm/12slm1108.png')

# initialize model
config_file = '/content/mmclassification/configs/resnet/resnet34_8xb32_in1k.py'
checkpoint_file = '/content/mmclassification/work_dirs/resnet34_8xb32_in1k/latest.pth'
model = init_model(config_file, checkpoint_file, device='cuda')

# perform inference
result = inference_model(model, img)
print(result)

how can i fix?? please tell me the sloution..

**Comment by Ezra-Yu** [comment]
modify this section code https://github.com/open-mmlab/mmclassification/blob/fb16bdc6a25f96843f8fb97b02e45872c61dc566/mmcls/apis/inference.py#L84-L90

To:
```
    with torch.no_grad():
        scores = model(return_loss=False, **data)
        pred_score = np.max(scores, axis=1)[0]
        pred_label = np.argmax(scores, axis=1)[0]
        result = {'pred_label': pred_label, 'pred_score': float(pred_score), 'scores': scores} # scores is a torch.Tensor. 
    result['pred_class'] = model.CLASSES[result['pred_label']]
    return result
```

**Reply by sinyoung-park** [reply]
thanks for your response. but how should I modify it?

**Reply by Ezra-Yu** [reply]
maybe you can cp the code of `inference_model`. and then rename it as `my_inference_model` to do the modification and use the new `my_inference_model`.

**Reply by sinyoung-park** [reply]
> maybe you can cp the code of `inference_model`. and then rename it as `my_inference_model` to do the modification and use the new `my_inference_model`.

i made my_inference_model. and how should i modify  this code?

with torch.no_grad():
        scores = model(return_loss=False, **data)
        pred_score = np.max(scores, axis=1)[0]
        pred_label = np.argmax(scores, axis=1)[0]
        result = {'pred_label': pred_label, 'pred_score': float(pred_score), 'scores': scores} # scores is a torch.Tensor. 
    result['pred_class'] = model.CLASSES[result['pred_label']]
    return result

please give me the code. I'm still not good enough at studying code. sorry

**Reply by Ezra-Yu** [reply]
modify the code in https://github.com/open-mmlab/mmclassification/blob/master/demo/image_demo.py 
```
# Copyright (c) OpenMMLab. All rights reserved.
from argparse import ArgumentParser

import mmcv

from mmcls.apis import my_inference_model, init_model, show_result_pyplot

def my_inference_model(model, img):
    """Inference image(s) with the classifier.
    Args:
        model (nn.Module): The loaded classifier.
        img (str/ndarray): The image filename or loaded image.
    Returns:
        result (dict): The classification results that contains
            `class_name`, `pred_label` and `pred_score`.
    """
    cfg = model.cfg
    device = next(model.parameters()).device  # model device
    # build the data pipeline
    if isinstance(img, str):
        if cfg.data.test.pipeline[0]['type'] != 'LoadImageFromFile':
            cfg.data.test.pipeline.insert(0, dict(type='LoadImageFromFile'))
        data = dict(img_info=dict(filename=img), img_prefix=None)
    else:
        if cfg.data.test.pipeline[0]['type'] == 'LoadImageFromFile':
            cfg.data.test.pipeline.pop(0)
        data = dict(img=img)
    test_pipeline = Compose(cfg.data.test.pipeline)
    data = test_pipeline(data)
    data = collate([data], samples_per_gpu=1)
    if next(model.parameters()).is_cuda:
        # scatter to specified GPU
        data = scatter(data, [device])[0]

    # forward the model
    with torch.no_grad():
        scores = model(return_loss=False, **data)
        pred_score = np.max(scores, axis=1)[0]
        pred_label = np.argmax(scores, axis=1)[0]
        result = {'pred_label': pred_label, 'pred_score': float(pred_score),  'scores': scores}
    result['pred_class'] = model.CLASSES[result['pred_label']]
    return result

def main():
    parser = ArgumentParser()
    parser.add_argument('img', help='Image file')
    parser.add_argument('config', help='Config file')
    parser.add_argument('checkpoint', help='Checkpoint file')
    parser.add_argument(
        '--show',
        action='store_true',
        help='Whether to show the predict results by matplotlib.')
    parser.add_argument(
        '--device', default='cuda:0', help='Device used for inference')
    args = parser.parse_args()

    # build the model from a config file and a checkpoint file
    model = init_model(args.config, args.checkpoint, device=args.device)
    # test a single image
    result = my_inference_model(model, args.img)
    # show the results
    print(mmcv.dump(result, file_format='json', indent=4))
    scores = result['scores']
```

use the code as 
```
python demo/image_demo.py ${IMG_PATH} ${CONFIG} ${CHECK_POINT}
```

**Reply by sinyoung-park** [reply]
> modify the code in https://github.com/open-mmlab/mmclassification/blob/master/demo/image_demo.py
> 
> ```
> # Copyright (c) OpenMMLab. All rights reserved.
> from argparse import ArgumentParser
> 
> import mmcv
> 
> from mmcls.apis import my_inference_model, init_model, show_result_pyplot
> 
> def my_inference_model(model, img):
>     """Inference image(s) with the classifier.
>     Args:
>         model (nn.Module): The loaded classifier.
>         img (str/ndarray): The image filename or loaded image.
>     Returns:
>         result (dict): The classification results that contains
>             `class_name`, `pred_label` and `pred_score`.
>     """
>     cfg = model.cfg
>     device = next(model.parameters()).device  # model device
>     # build the data pipeline
>     if isinstance(img, str):
>         if cfg.data.test.pipeline[0]['type'] != 'LoadImageFromFile':
>             cfg.data.test.pipeline.insert(0, dict(type='LoadImageFromFile'))
>         data = dict(img_info=dict(filename=img), img_prefix=None)
>     else:
>         if cfg.data.test.pipeline[0]['type'] == 'LoadImageFromFile':
>             cfg.data.test.pipeline.pop(0)
>         data = dict(img=img)
>     test_pipeline = Compose(cfg.data.test.pipeline)
>     data = test_pipeline(data)
>     data = collate([data], samples_per_gpu=1)
>     if next(model.parameters()).is_cuda:
>         # scatter to specified GPU
>         data = scatter(data, [device])[0]
> 
>     # forward the model
>     with torch.no_grad():
>         scores = model(return_loss=False, **data)
>         pred_score = np.max(scores, axis=1)[0]
>         pred_label = np.argmax(scores, axis=1)[0]
>         result = {'pred_label': pred_label, 'pred_score': float(pred_score),  'scores': scores}
>     result['pred_class'] = model.CLASSES[result['pred_label']]
>     return result
> 
> def main():
>     parser = ArgumentParser()
>     parser.add_argument('img', help='Image file')
>     parser.add_argument('config', help='Config file')
>     parser.add_argument('checkpoint', help='Checkpoint file')
>     parser.add_argument(
>         '--show',
>         action='store_true',
>         help='Whether to show the predict results by matplotlib.')
>     parser.add_argument(
>         '--device', default='cuda:0', help='Device used for inference')
>     args = parser.parse_args()
> 
>     # build the model from a config file and a checkpoint file
>     model = init_model(args.config, args.checkpoint, device=args.device)
>     # test a single image
>     result = my_inference_model(model, args.img)
>     # show the results
>     print(mmcv.dump(result, file_format='json', indent=4))
>     scores = result['scores']
> ```
> 
> use the code as
> 
> ```
> python demo/image_demo.py ${IMG_PATH} ${CONFIG} ${CHECK_POINT}
> ```

es. There seems to be a misunderstanding. I was asking how to modify this model. I created my_inference.py via %%writfile and have been copying the contents from inference.py. But I need to modify the #forward part. I would be grateful if you could tell me how to modify the #forward.

**Reply by Ezra-Yu** [reply]
why need to modify the `model.forward`?

**Reply by sinyoung-park** [reply]
> why need to modify the `model.forward`?

reason is that you said to fix the forward section in your reply above. 
and i want return to score of each class. that is my purpose.
so i modify the code. the inference.py is only return to single score. 
appreciate your help, but I wish you could do more.

**Reply by Ezra-Yu** [reply]
`model.forward` has return  the scores of each class, but we only return the final score for the predict class in function `inference_model`

this statement
```
scores = model(return_loss=False, **data) # the scores are the scores for all the classes
```

**Reply by sinyoung-park** [reply]
> `model.forward` has return the scores of each class, but we only return the final score for the predict class in function `inference_model`
> 
> this statement
> 
> ```
> scores = model(return_loss=False, **data) # the scores are the scores for all the classes
> ```

ok. and than, how to print score? i do print(scores), the return is not define...

**Reply by Ezra-Yu** [reply]
new a `pred.py` file and cp the following code into the file
```
# Copyright (c) OpenMMLab. All rights reserved.
from argparse import ArgumentParser

import mmcv

from mmcls.apis import my_inference_model, init_model, show_result_pyplot

def my_inference_model(model, img):
    """Inference image(s) with the classifier.
    Args:
        model (nn.Module): The loaded classifier.
        img (str/ndarray): The image filename or loaded image.
    Returns:
        result (dict): The classification results that contains
            `class_name`, `pred_label` and `pred_score`.
    """
    cfg = model.cfg
    device = next(model.parameters()).device  # model device
    # build the data pipeline
    if isinstance(img, str):
        if cfg.data.test.pipeline[0]['type'] != 'LoadImageFromFile':
            cfg.data.test.pipeline.insert(0, dict(type='LoadImageFromFile'))
        data = dict(img_info=dict(filename=img), img_prefix=None)
    else:
        if cfg.data.test.pipeline[0]['type'] == 'LoadImageFromFile':
            cfg.data.test.pipeline.pop(0)
        data = dict(img=img)
    test_pipeline = Compose(cfg.data.test.pipeline)
    data = test_pipeline(data)
    data = collate([data], samples_per_gpu=1)
    if next(model.parameters()).is_cuda:
        # scatter to specified GPU
        data = scatter(data, [device])[0]

    # forward the model
    with torch.no_grad():
        scores = model(return_loss=False, **data)
        pred_score = np.max(scores, axis=1)[0]
        pred_label = np.argmax(scores, axis=1)[0]
        result = {'pred_label': pred_label, 'pred_score': float(pred_score),  'scores': scores}
    result['pred_class'] = model.CLASSES[result['pred_label']]
    return result

def main():
    parser = ArgumentParser()
    parser.add_argument('img', help='Image file')
    parser.add_argument('config', help='Config file')
    parser.add_argument('checkpoint', help='Checkpoint file')
    parser.add_argument(
        '--show',
        action='store_true',
        help='Whether to show the predict results by matplotlib.')
    parser.add_argument(
        '--device', default='cuda:0', help='Device used for inference')
    args = parser.parse_args()

    # build the model from a config file and a checkpoint file
    model = init_model(args.config, args.checkpoint, device=args.device)
    # test a single image
    result = my_inference_model(model, args.img)
    # show the results
    print(mmcv.dump(result, file_format='json', indent=4))
    scores = result['scores']
```

than run :
```
python pred.py ${YOUR_IMG_PATH} ${YOUR_CONFIG} ${YOUR_CHECK_POINT}
```

**Reply by Ezra-Yu** [reply]
I hope you can understand the code in the script

**Reply by sinyoung-park** [reply]
i run this model, and i do print(result) return is {'pred_label': 1, 'pred_score': 0.8242934942245483, 'pred_class': '12slm'} it is only one class...

**Reply by Ezra-Yu** [reply]
It seems you are not running the script I give to you.

Maybe you can try the [`1.x` branch](https://github.com/open-mmlab/mmclassification/tree/dev-1.x). refer to this [tutorial](https://mmclassification.readthedocs.io/en/dev-1.x/user_guides/inference.html). which need not modify the code.

**Reply by sinyoung-park** [reply]
> It seems you are not running the script I give to you.
> 
> Maybe you can try the [`1.x` branch](https://github.com/open-mmlab/mmclassification/tree/dev-1.x). refer to this [tutorial](https://mmclassification.readthedocs.io/en/dev-1.x/user_guides/inference.html). which need not modify the code.

Your code is up and running successfully! Thank you for your help! The problem with this code seems to be a naming error: my_inference_model and inference_model were misnamed. I replaced my_inference_model with infer_image and typed print('Scores',result['scores']) and got the desired result! Thanks for the kind explanation.
